robottelo/decorators/__init__.py

```diff
@@ -368,3 +369,25 @@ def decorator(func):
         )(func)
 
     return decorator
+
+
+def report_if_failed(func):
+    """report test status in global_dict if it is failed. This will be helpful
+    for skipping post_upgrade tests.
+
+    Decorating a method::
+
+        @report_if_failed
+        def test_pre_upgrade(self):
+            assert 0
+    """
+    def wrapper(*args, **kwargs):
+        try:
+            func(*args, **kwargs)
+        except Exception:
+            status_dict = {
+                args[0].__class__.__name__:
+                    {'{0}'.format(func.__name__): "failed"}}
```

Comment:
can you look for function get_func_name in robottelo.helpers , this should help create a function name for different situations.

Reply:
also take a look at function https://github.com/SatelliteQE/robottelo/blob/master/tests/foreman/conftest.py#L163
that may filter tests after collection

Reply:
thanks @ldjebran 
will work on it

Reply:
@omkarkhatavkar @san7ket @jyejare @ntkathole I suggest to switch to pytest 

and use all of this, eg passing functions as arguments instead of functions names that will be more adequate and not have to deal with test names , test names will be returned and compared at low level. 

```python

@pre_upgrade
def test_capsule_pre_upgrade_1():
    pass

@pre_upgrade
def test_capsule_pre_upgrade_2():
    pass

@post_upgrade(depend_on=[test_capsule_pre_upgrade_1, test_capsule_pre_upgrade_2])
def test_capsule_post_upgrade():
    pass





```
What do you think?

Reply:
Agree with you

Reply:
@omkarkhatavkar @san7ket @jyejare @ntkathole
After many instigation come to this solution for pytest:
consider the following conftest.py
```python 
import json
import logging
import os

from pytest import hookspec

from robottelo.helpers import get_func_name

LOGGER = logging.getLogger('robottelo')

pre_upgrade_failed_tests = []


PRE_UPGRADE_TESTS_FILE_OPTION = 'pre_upgrade_tests_file'
PRE_UPGRADE_TESTS_FILE_PATH = '/var/tmp/pre_upgrade_tests_file.json'
PRE_UPGRADE = False
POST_UPGRADE = False


def pytest_addoption(parser):
    parser.addoption(
        '--{0}'.format(PRE_UPGRADE_TESTS_FILE_OPTION),
        action='store',
        default=PRE_UPGRADE_TESTS_FILE_PATH
    )


def pytest_report_header(config):
    global pre_upgrade_failed_tests
    global PRE_UPGRADE
    global POST_UPGRADE
    global PRE_UPGRADE_TESTS_FILE_PATH
    PRE_UPGRADE_TESTS_FILE_PATH = getattr(config.option, PRE_UPGRADE_TESTS_FILE_OPTION)
    if 'pre_upgrade' in config.option.markexpr:
        pre_upgrade_failed_tests = []
        PRE_UPGRADE = True
        # remove file before begin
        if os.path.exists(PRE_UPGRADE_TESTS_FILE_PATH):
            os.unlink(PRE_UPGRADE_TESTS_FILE_PATH)
    if 'post_upgrade' in config.option.markexpr:
        if PRE_UPGRADE:
            raise Exception('options error: cannot do pre_upgrade and post_upgrade at the same'
                            ' time')
        POST_UPGRADE = True
        pre_upgrade_failed_tests = []
        if os.path.exists(PRE_UPGRADE_TESTS_FILE_PATH):
            with open(PRE_UPGRADE_TESTS_FILE_PATH, 'r') as json_file:
                pre_upgrade_failed_tests = json.load(json_file)


@hookspec(firstresult=True)
def pytest_report_teststatus(report):
    global pre_upgrade_failed_tests
    if report.outcome != 'passed':
        pre_upgrade_failed_tests.append(report.nodeid)


def pytest_sessionfinish(session, exitstatus):
    """called after whole test run finished, right before returning the exit status to the system.

    :param _pytest.main.Session session: the pytest session object
    :param int exitstatus: the status which pytest will return to the system
    """
    global pre_upgrade_failed_tests
    if PRE_UPGRADE and exitstatus:
        pre_upgrade_items = [
            item
            for item in session.items if item.get_marker('pre_upgrade') is not None
        ]
        pre_upgrade_tests_to_report = [
            get_func_name(item.function, item)
            for item in pre_upgrade_items if item.nodeid in pre_upgrade_failed_tests
        ]
        LOGGER.info('Save failed tests to file %s', PRE_UPGRADE_TESTS_FILE_PATH)
        with open(PRE_UPGRADE_TESTS_FILE_PATH, 'w') as json_file:
            json.dump(pre_upgrade_tests_to_report, json_file)


def pytest_collection_modifyitems(items, config):
    """ called after collection has been performed, will skip item/post-upgrade test if
    pre-upgrade test status is failed.
    """
    post_upgrade_items = [
        item
        for item in items if item.get_marker('post_upgrade') is not None
    ]
    deselected_items = []
    for item in post_upgrade_items:
        dependant_on_functions = []
        for marker in item.iter_markers('post_upgrade'):
            depend_on = marker.kwargs.get('depend_on')
            if isinstance(depend_on, (list, tuple)):
                dependant_on_functions.extend(depend_on)
            else:
                dependant_on_functions.append(depend_on)
            depend_on_names = [get_func_name(f) for f in dependant_on_functions]
            for depend_on_name in depend_on_names:
                if depend_on_name in pre_upgrade_failed_tests:
                    LOGGER.info('Deselected (because of dependant test failed): %s',
                                get_func_name(item.function, item))
                    deselected_items.append(item)

    config.hook.pytest_deselected(items=deselected_items)
    items[:] = [item for item in items if item not in deselected_items]
```
and consider the test_upgrade.py
```python
import pytest

pre_upgrade = pytest.mark.pre_upgrade
post_upgrade = pytest.mark.post_upgrade


@pre_upgrade
def test_capsule_pre_upgrade_1():
    pass


@pre_upgrade
def test_capsule_pre_upgrade_2():
    assert 1 == 0


@post_upgrade(depend_on=[test_capsule_pre_upgrade_1, test_capsule_pre_upgrade_2])
def test_capsule_post_upgrade_1():
    pass


@post_upgrade(depend_on=[test_capsule_pre_upgrade_1])
def test_capsule_post_upgrade_2():
     pass

```
running pre_upgrade tests
```
(robottelo) dlezz@dlezz:~/projects/robottelo-fork$ pytest -v -m 'pre_upgrade' tests/foreman/draft/test_upgrade.py 
===================================================================================================== test session starts =====================================================================================================
platform linux -- Python 3.6.6, pytest-4.0.1, py-1.5.4, pluggy-0.8.0 -- /home/dlezz/.pyenv/versions/robottelo/bin/python3.6
cachedir: .pytest_cache
shared_function enabled - ON - scope: ffff - storage: redis
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: wait-for-1.0.9, xdist-1.22.5, services-1.3.1, mock-1.10.0, forked-0.2, cov-2.5.1
collecting ... 2018-12-10 17:25:28 - conftest - DEBUG - BZ deselect is disabled in settings

collected 4 items / 2 deselected                                                                                                                                                                                              

tests/foreman/draft/test_upgrade.py::test_capsule_pre_upgrade_1 PASSED                                                                                                                                                  [ 50%]
tests/foreman/draft/test_upgrade.py::test_capsule_pre_upgrade_2 FAILED                                                                                                                                                  [100%]

========================================================================================================== FAILURES ===========================================================================================================
_________________________________________________________________________________________________ test_capsule_pre_upgrade_2 __________________________________________________________________________________________________

    @pre_upgrade
    def test_capsule_pre_upgrade_2():
>       assert 1 == 0
E       assert 1 == 0

tests/foreman/draft/test_upgrade.py:14: AssertionError
---------------------------------------------------------------------------------------------------- Captured stdout setup ----------------------------------------------------------------------------------------------------
2018-12-10 17:25:28 - robottelo - DEBUG - Started Test: test_upgrade.py/test_capsule_pre_upgrade_2
----------------------------------------------------------------------------------------------------- Captured log setup ------------------------------------------------------------------------------------------------------
conftest.py                135 DEBUG    Started Test: test_upgrade.py/test_capsule_pre_upgrade_2
-------------------------------------------------------------------------------------------------- Captured stdout teardown ---------------------------------------------------------------------------------------------------
2018-12-10 17:25:28 - robottelo - DEBUG - Finished Test: test_upgrade.py/test_capsule_pre_upgrade_2
---------------------------------------------------------------------------------------------------- Captured log teardown ----------------------------------------------------------------------------------------------------
conftest.py                137 DEBUG    Finished Test: test_upgrade.py/test_capsule_pre_upgrade_2
====================================================================================== 1 failed, 1 passed, 2 deselected in 0.13 seconds =======================================================================================
2018-12-10 17:25:28 - robottelo - INFO - Save failed tests to file /var/tmp/pre_upgrade_tests_file.json
(robottelo) dlezz@dlezz:~/projects/robottelo-fork$ 
```
by default the file is saved to '/var/tmp/pre_upgrade_tests_file.json' but as option you can pass -- --pre_upgrade_tests_file "/var/tmp/failed_tests.json"
**Notice the line: 2018-12-10 17:25:28 - robottelo - INFO - Save failed tests to file /var/tmp/pre_upgrade_tests_file.json** 


running the post_upgrade tests:
```
(robottelo) dlezz@dlezz:~/projects/robottelo-fork$ pytest -v -m 'post_upgrade' tests/foreman/draft/test_upgrade.py 
===================================================================================================== test session starts =====================================================================================================
platform linux -- Python 3.6.6, pytest-4.0.1, py-1.5.4, pluggy-0.8.0 -- /home/dlezz/.pyenv/versions/robottelo/bin/python3.6
cachedir: .pytest_cache
shared_function enabled - ON - scope: ffff - storage: redis
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: wait-for-1.0.9, xdist-1.22.5, services-1.3.1, mock-1.10.0, forked-0.2, cov-2.5.1
collecting ... 2018-12-10 17:33:18 - conftest - INFO - Deselected (because of dependant test failed): tests.foreman.draft.test_upgrade.test_capsule_post_upgrade_1

2018-12-10 17:33:18 - conftest - DEBUG - BZ deselect is disabled in settings

collected 4 items / 3 deselected                                                                                                                                                                                              

tests/foreman/draft/test_upgrade.py::test_capsule_post_upgrade_2 PASSED                                                                                                                                                 [100%]

=========================================================================================== 1 passed, 3 deselected in 0.01 seconds ============================================================================================
```
**Notice the line: collecting ... 2018-12-10 17:33:18 - conftest - INFO - Deselected (because of dependant test failed): tests.foreman.draft.test_upgrade.test_capsule_post_upgrade_1**